### ISSUE TITLE
No Directive annotation found on MdAnchor

### ISSUE THREAD

**Issue by gstrit** [justindujardin/ng2-material#38]
Hello,

I'm trying to integrate ng2-material to my project but I have this error message : 
EXCEPTION: No Directive annotation found on MdAnchor

Thanks

**Comment by ollwenjones** [comment]
@gstrit what version of angular2 are you using? This might have to do with api changes between beta.0 and .1 & .2

**Comment by alalonde** [comment]
Looks light you're right: I am using beta.2, downgrading to beta.0 fixed it.

**Comment by gstrit** [comment]
Yes i'm using beta.2. I will try with beta.0
Thanks

**Comment by ollwenjones** [comment]
Spending some time on upgrading my fork to beta.3 this morning, since I'm not real interested in rolling my other project back, atm. Wish me luck.

**Comment by gstrit** [comment]
Still the same issue with beta.0

**Comment by ollwenjones** [comment]
Stuff upgraded and working in my fork's beta3 branch, but I won't have time to sort out the remaining unit test issues today. I'm interested to try it out in the project I'm working in and see what I get.

**Comment by ollwenjones** [comment]
@gstrit are you using webpack in your parent project, by chance?

**Comment by ollwenjones** [comment]
I'm making progress on the project I'm using these in (which uses beta.2) after importing the grunt build of my local repo (which is upgraded to beta.3). There are still Unit test issues, but I'm not likely to have time to work on those until I catch up on my real project.

My fork is here
but the Beta.3 upgrade should fix [this issue](https://github.com/justindujardin/ng2-material/issues/38),
as well as [41](https://github.com/justindujardin/ng2-material/issues/41)

**Comment by opsxcq** [comment]
+1 waiting for porting it to beta 3, since zone bug was only fixed in that angular version.

**Comment by justindujardin** [comment]
I published version 0.2.2 today that supports the latest Beta.3 build.  Please try it out and close this issue if your problems are resolved.

**Comment by ollwenjones** [comment]
0.2.2 didn't fix it for me... same error... `No Directive annotation found on MdAnchor` - which is pretty strange. My fork's beta 3 branch worked for me, and it was branched off your beta2 branch, which should have had the necessary fixes. (Most of the breaking changes from 2-3 were related to testing.)

**Comment by opsxcq** [comment]
@ollwenjones it fixed for me. You are using jspm or bower ?

**Comment by ollwenjones** [comment]
@opsxcq I had to update the parent project from beta2 to beta 3 - along with the zone 0.5.11 upgrade.
After that I still had the error and npm uninstall / npm install ng-material again and it works fine.
I think perhaps uninstalling and reinstalling ng-material was the crucial step... either way it's working for me now.

**Comment by opsxcq** [comment]
I'm using JSPM, same thing, had to remove then add it again to work.

**Comment by ollwenjones** [comment]
Big thanks to @justindujardin for the update. :+1:

**Comment by bvmCoder** [comment]
Hello Justin I need a help to link the ng2-material to my angular 2 app, I am getting following Error, also I am using WebPack

BaseException {message: "No Directive annotation found on MdAnchor", stack: "Error: No Directive annotation found on MdAnchor …000/vendor.bundle.js)

EXCEPTION: No Directive annotation found on MdAnchor


Have tried npm uninstall ng2-material and then reinstall using npm install ng2-material

**Comment by ollwenjones** [comment]
@bvmCoder, what version of angular2 beta is installed in your project?

**Comment by bvmCoder** [comment]
Thank you @ollwenjones (Sam Jones)

I have installed following: Beta Version 3

which is right here from my Package.json file

I am getting some webSocket pending status in my network Tab

"angular2": "^2.0.0-beta.3",

**Comment by ollwenjones** [comment]
@bvmCoder I think I just deleted the node_modules/ng2-material directory (versus npm install) to get it working. Beta 3 should be working, making sure you `npm install -S ng2-material@0.2.2`

**Comment by bvmCoder** [comment]
Thanks Sam @ollwenjones 
Quick Question what does npm install -S 
{{ -S stands for?}}

Thanks

**Comment by ollwenjones** [comment]
@bvmCoder `-S` is just shorthand for `--save` to "dependencies" `-D` is for `--save-dev` for dev dependencies. You can even do `npm i -D`

**Comment by bvmCoder** [comment]
@ollwenjones , Ohh okay thanks for the knowledge, as It is still not working, I am investigating with webSoket, as at my work place I think I have webSoket restriction. I am not sure what is the realistic problem.

**Comment by schippie** [comment]
@ollwenjones @justindujardin I am unsure but within beta6 of angular 2 the problem has returned. 
I have verified that the package.json version is 0.2.2 of ng2-material. But its throwing all kinds of errors:

Example repository:
https://bitbucket.org/schippie/webpack-ng2-material/overview

1.  git clone
2. npm install
3. npm start

**Comment by ollwenjones** [comment]
@schippie I have actually held off of the beta6 upgrade in my project in case something like this comes up, since I really need to get some tasking done before I have any time for upgrade churn, so I can't tell you for sure if beta6 is a culprit here or not.

I can share this:
```
"dependencies": {
  "angular2": "2.0.0-beta.3",
  "es6-promise": "3.0.2",
  "es6-shim": "0.33.3",
  "es7-reflect-metadata": "^1.4.0",
  "moment": "^2.11.0",
  "ng2-component-library": "git+ssh://git@bitbucket.org/openroads/ng2-component-library.git#8bc53cf",
  "ng2-material": "0.2.2",
  "rxjs": "5.0.0-beta.0",
  "zone.js": "0.5.11"
},
```
I _really_ thought zone 0.5.11 was a specific peer dep version... I poked around your repo, and I didn't see anything amiss. My only other main difference is a 'vendor' webpack entry point in addition to the 'main', but ng2-material wasn't imported there. Sorry I don't have time to investigate further atm.

**Comment by ollwenjones** [comment]
As for your 'Edit' I think it's really ideal in a webpack context to import library source files... in which case the dependencies of the components library don't necessarily come in to play, do they?

**Comment by dsmrs** [comment]
Hi! I try with your configuration, and it's not working :disappointed:  I still have the same `No Directive annotation found on MdAnchor` (Or other DirectiveName when I try to import only the directive I want). Like @schippie I started my project with the webpack-starter https://github.com/AngularClass/angular2-webpack-starter.git

**Comment by schippie** [comment]
@dsmrs @ollwenjones Based on help from gitter we did manage to figure out how to get it work by copying over the source directory into your own project. This fixes the bugs i was having as well.

Though when trying to use the source folder directly through:

ng2-material/source/all (for example) 

It still throws the same exception. It almost seems as if there are 2 different injectors which cause the services not to be available to the directives (when you use the ng2-material/source/all).

**Comment by ollwenjones** [comment]
@dsmrs I used the same Angular Class starter project, though I've modified the configuration somewhat since. 
@schippie I wish I had time to investigate this more... be a bummer to have to copy source directory in. The injectors in ng2 are tiered by context, from what I understand, instead of being global singleton as in ng1 (don't think I could explain it very well, though.)

There is a question in my mind about whether this might be related to api changes though, regarding @view annotation and what it was for and whether people really needed to add it all the time, that is perhaps making this more brittle.

This is going to bug me. We'll all have a better understanding of ng2 and webpack if we can nail this down.

**Comment by ollwenjones** [comment]
@schippie I noticed your [tsconfig.json](https://bitbucket.org/schippie/webpack-ng2-material/src/bbb89737765f7499f14b6ba13f5910a9b002247c/tsconfig.json?fileviewer=file-view-default) doesn't seem to be importing any typings? Could that be related / am I missing something?

About the two injectors: I'm pretty sure that _is_ the case. When you add ng2-material via `CUSTOM_PROVIDERS` in main.ts, it creates one global instance, but when you add it to the providers array as `MATERIAL_DIRECTIVES` it creates another for that part of the app tree. At least I've noticed that with services, adding them globally, but also to a component providers array creates duplicates. I think the `MATERIAL_DIRECTIVES` import should be in the `directives` property though, not the `providers`.

**Comment by schippie** [comment]
@ollwenjones Sorry about the mistake within the providers I was experimenting with a lot of stuff within this project to get it to work and before I committed it I probably put it back in the wrong way (corrected it in the last commit)

But normally i should not have to add the providers again to the component directive right? I can experiment what would happen if I add the services to the individual components though

**Comment by schippie** [comment]
@ollwenjones  I HAVE FIGURED IT OUT :tongue: 

The problem is the fact that within the ng2-material directive after my npm-shrinkwrap trick (to make the package compatible with angular beta 6) a node_modules folder is located. this node_modules folder contains another angular2 package (which is the forced beta 6 package) what happens now is that when including the source of ng2-material in the following manner:

ng2-material/all

That it grabs that angular2 package in its own requires instead. Which probably leads to some duplication problems which leads to all the weird errors. 
So i think the only thing this package has to do is have its package.json version changed to angular beta 7 (and everything should work on installation again).

**Comment by justindujardin** [comment]
This seems to be a problem for a number of people, could one of you please provide clear steps to reproduce it, so we can look at a fix?

**Comment by schippie** [comment]
1. Install angular2@beta6 or higher in combination with ng2-material 
2. Try to run it through the typescript compiler this at this point will throw a lot of duplication errors. 
3. To work around this problem I used npm-shrinkwrap to force the version that ng2-material uses as dependencies (angular2/zonejs etc.) to be the same as angular2@beta6.This lead to the error described above `No Directive annotation found on MdAnchor`.
4. After this point you start testing a lot of different ways to fix the above error as it turns out the simplest fix was to simple rename the node_modules folder within the ng2-materials folder directive to  _node_modules. 

And I think the reason this fixes the issues is that it no longer includes the wrong versions or creates duplicate versions of the zone.js / angular2 etc. 

If you want you can test it using this repository (which i have already done steps 1, 2 and 3):
https://bitbucket.org/schippie/webpack-ng2-material/src 

Though since you have already merged https://github.com/justindujardin/ng2-material/pull/66 I think the issues should be resolved for all users. 
Going to try it out and reply back here with if it is fixed for me when doing a clean install :smile:

**Comment by justindujardin** [comment]
@schippie this should be fixed in 0.2.5 release. I cloned your project and verified it works for me with the latest build.  I'm going to close this for now.

**Comment by bvmCoder** [comment]
@ollwenjones  we are still using Angular 2 Beta 3 and we have planned to upgrade to Beta 6, but currently I am working on the prototype app to user Redux/React

**Comment by msegers** [comment]
Strange, I'm also using webpack, I had the same issue. Reinstalling fixed it for me.

**Comment by rsnider19** [comment]
The solution @schippie provided works for Beta.7 and ng2-material 0.2.10 with webpack

**Comment by wmaurer** [comment]
I've just experienced this error using webpack, and I think it really has to do with the fact that I had two different versions of angular2 installed.

My `package.json` has `"angular2": "^2.0.0-beta.8"`.
ng2-material's `package.json` has `"angular2": "2.0.0-beta.8"`

Because `beta.9` got released yesterday, the root of my `node_modules` has `beta.9`, and a second version `beta.8` is installed under `node_modules/ng2-material/node_modules/angular2`.
I believe this is the source of the problem.

My suggestion would be to remove **all** `dependencies` from the `ng2-material` `package.json`.
And instead of having `dependencies`, have `peerDependencies` as follows: 
```
  "peerDependencies": {
    "angular2": "^2.0.0-beta.7",
    "rxjs": "^5.0.0-beta.2"
  },
```

I have seen this setup in a number of other libraries that depend on ng2, e.g. [ngrx](https://github.com/ngrx/store/blob/master/package.json) (written by a ng2 team member).

I think this is especially important at the moment, since a new ng2 beta is released every week. What do you think @justindujardin?

**Comment by schippie** [comment]
@wmaurer I think indeed this would once and for all fix the issues experienced for people here. If we are talking about modifications to the package.json i would also like to suggest moving the typescript dependency into the dev dependencies (or would that require a seperate issue to be opened @justindujardin  ? ).

**Comment by wmaurer** [comment]
@schippie Agree, both typescript and systemjs should be moved into devDependencies.

@justindujardin would you accept a PR with these changes?

**Comment by ekarlso** [comment]
I am getting hit the the MdAnchor bug as well on beta9, why is ithis happening ?

**Comment by schippie** [comment]
@ekarlso If you are using webpack as your development tool to compile typescript my assumption is as follows:

You npm install the packages. If the version of angular2 does not match the version of a dependency (who also has a dependency on angular2) it causes  npm to put that in a sub-folder of the dependency (which is correct). 

So the structure becomes something like:

```
|->node_modules
|      |-> angular2@beta9
|      |->ng2-material
|      |            |->angular2@beta8
```

What happens now is that during the bundling, your typescript is compiled to javascript. The `import` statements change to `require()`. 

Ng2-materials `import {something something} from 'angular2/something` grabs the sub package within ng2-material according to the commonJS logic (Which is correct). 

The idea here is that a package can have a dependency on a different version (one of the benefits of npm packages). But with angular2 it causes some weird duplication error that is not shown to the developer. 

If you are suffering from this problem currently the easiest fix is to go into the ng2-material sub-folder and rename the `node_modules` folder to `_node_modules` until the dependencies within the ng2-material package.json are changed to peer dependencies.

**Comment by ollwenjones** [comment]
Great thorough explanation @schippie

**Comment by justindujardin** [comment]
@wmaurer thanks for the suggestion.  I have done three things:

1. move system.js and typescript into devDependencies
2. Use ^ version matches on angular2, rxjs, and zone.js
3. Assign peerDependencies = dependencies in published npm package.

The latest version 0.2.11 contains these changes.  Hopefully this will resolve the issues people are having here.  If there's more to do that I missed, please open a PR or suggest it here.

**Comment by wmaurer** [comment]
Thanks for the fix @justindujardin 
Now, because of ^ versions, it should work (hard to test right now as both my app and ng2-material specifiy the same ng2 version)

I'm not completely sure on `es6-promise`, `es6-shim`, `reflect-metadata`, `zone.js` being in the `peerDependencies`, especially those which have exact versions specified.
These are peerDependencies of angular2, and I would let the angular2 package specify their versions. It's not causing any problem right now, but it might in the future ...

**Comment by MilosStanic** [comment]
Still having this issue with Angular2-beta.7

Note: I'm using angular2-meteor, and it has some importing issues for which I had to import ng2-material via require. But I don't think that's the issue here.

I have been trying to repackage ng2-material for use with MeteorJS via their packaging system Atmosphere, but to no avail.

**Comment by wmaurer** [comment]
@MilosStanic that's because @justindujardin has set the minimum supported version (via peerDependencies) to angular2-beta.9
I would suggest upgrading to beta.9 or otherwise you'll have to convince @justindujardin to support old beta versions

**Comment by MilosStanic** [comment]
@wmaurer oh, okay. I'll have to wait until angular2-meteors upgrades to angular2-beta.9 :/
Thanks. I don't want to convince Justin of anything. Always better to move ahead.
